File: acceptance-tests/src/test/java/com/timetracker/StepDefs.java
package com.timetracker;

import io.cucumber.java.Before;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;

import java.util.Map;

import static java.lang.Runtime.getRuntime;
import static java.lang.System.getProperty;
import static java.lang.System.setProperty;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class StepDefs {
    private static final String CHROME_DRIVER_PATH = "src/test/resources/chrome-drivers/";
    private static final Map<String, String> chromeDrivers = Map.of(
            "Linux", "chromedriver_linux",
            "Mac OS X", "chromedriver_mac",
            "Windows", "chromedriver.exe" // TODO Verify map key on a Windows 10 machine
    );
    public static WebDriver driver;

    private static String getChromeDriver() {
        return CHROME_DRIVER_PATH + chromeDrivers.getOrDefault(
                getProperty("os.name"),
                "OS Not Supported. Add chrome driver to " + CHROME_DRIVER_PATH + " and make an entry in chromeDrivers in this class."
        );
    }

    @Before
    public void initWebDriver() {
        if (driver != null) return;
        setProperty("webdriver.chrome.driver", getChromeDriver());
        driver = new ChromeDriver();
        getRuntime().addShutdownHook(new Thread(driver::quit));
    }

    @Given("I open the home page")
    public void open_home_page() {
        driver.get("http://localhost:3000");
    }

    @Then("I MUST see a message {string}")
    public void iMustSeeMessage(String message) {
        assertEquals(driver.findElement(By.className("App")).getText(), message);
    }
}

File: acceptance-tests/src/test/java/com/timetracker/StepDefs.kt
package com.timetracker

import io.cucumber.java.en.Given
import io.cucumber.java.en.Then
import org.junit.jupiter.api.Assertions.assertEquals
import org.openqa.selenium.By
import org.openqa.selenium.WebDriver
import org.openqa.selenium.chrome.ChromeDriver
import org.openqa.selenium.chrome.ChromeOptions
import org.openqa.selenium.remote.RemoteWebDriver

class StepDefs {
    @Given("I open the home page")
    fun open_home_page() {
        driver["http://localhost:3000"]
    }

    @Then("I MUST see a message {string}")
    fun iMustSeeMessage(message: String?) {
        assertEquals(driver.findElement(By.className("App")).text, message)
    }

    companion object {
        private const val CHROME_DRIVER_PATH = "src/test/resources/chrome-drivers/"
        private val chromeDrivers = mapOf(
                "Linux" to "chromedriver_linux",
                "Mac OS X" to "chromedriver_mac",
                "Windows" to "chromedriver.exe" // TODO Verify map key on a Windows 10 machine
        )

        val driver: WebDriver = run {
            System.setProperty("webdriver.chrome.driver", chromeDriver)
            val options = ChromeOptions()
            options.addArguments("--headless")
            Runtime.getRuntime().addShutdownHook(Thread(Runnable { (driver as RemoteWebDriver).quit() }))
            ChromeDriver(options)
        }

        private val chromeDriver: String
            get() = CHROME_DRIVER_PATH + chromeDrivers.getOrDefault(
                    System.getProperty("os.name"),
                    "OS Not Supported. Add chrome driver to $CHROME_DRIVER_PATH and make an entry in chromeDrivers in this class."
            )
    }
}
